<b>Привет, Евгений!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовил замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌ Замечание по всему разделу ответов на вопросы 
    
В каждом вопросе нужно построить сводную. Здесь можно воспользоваться `pivot_table()`. Попробуй получить сводную следующего вида. В индексе столбец, который исследуем (в первом вопросе это количество детей), а столбцы: 'Всего кредитополучателей', 'Всего должников', 'Доля должников'.

Также могу посоветовать переименовать столбцы и написать функцию для построения сводных во всех вопросах.
    <div class='alert alert-info'>Сделал функцию для всех вычислений (е стал делать параметр debt поумолчанию). В каждом случае применил pivot_table. Переименовал столбцы. </div>
</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
def ds_pivot(data, index, values):
    """Построение сводной таблицы по заданным критериям."""
    pivot = data.pivot_table(index=index, values=values, aggfunc=['count', 'sum'])
    pivot.columns = ['count', 'sum']
    pivot['duty'] = pivot['sum'] / pivot['count']
    pivot = pivot.rename(columns={'count': 'Всего', 'sum': 'Должников', 'duty': 'Доля_должников'})
    return pivot

In [23]:
debts_child = ds_pivot(data, 'children', 'debt')
print(debts_child)

          Всего  Должников  Доля_должников
children                                  
0         14091       1063        0.075438
1          4808        444        0.092346
2          2052        194        0.094542
3           330         27        0.081818
4            41          4        0.097561
5             9          0        0.000000


<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅ 
    
Для вывода датафрейма на экран можно использовать `display()` вместо `print()`
</div>

**Вывод:** 
Рассматриваем результаты для случаев с достаточным для анализа количеством данных - для 0, 1, 2 имеющихся детей.

<em>С появлением и увеличением количества детей доля должников по кредитам возрастает.</em>

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Выявлять зависимость стоит, сравнивая самые большие, сопоставимые по размеру категории заемщиков. Какую зависимость можно выявить, сравнив категории с 0, 1 и 2 детьми?
    <div class="alert alert-info">Внёс исправления</div>
</div>


<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Данных по многодетным (3 и более детей) недостаточно для корректных выводов по ним
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
debts_status = ds_pivot(data, 'family_status', 'debt')
print(debts_status)

                       Всего  Должников  Доля_должников
family_status                                          
Не женат / не замужем   2796        273        0.097639
в разводе               1189         84        0.070648
вдовец / вдова           951         63        0.066246
гражданский брак        4134        385        0.093130
женат / замужем        12261        927        0.075606


**Вывод:**
Одинокие люди, утратившие статус замужем/женат лучше выплачивают кредиты. Хужевсего выплачивают люди вообще не имеющие статуса.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Люди, побывавшие в официальном браке, наиболее надежны. Возможно, есть связь с возрастом
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
print(data['total_income'].max(), data['total_income'].min())  # Разброс зарплат.



2265604 20667


<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Категоризация уже была выше в предобработке в задании 16. Нужно убрать дублирование кода
    <div class="alert alert-info">
    Убрал повторение функции.
</div>
</div>

In [26]:
data['categorize_income'] = data['total_income'].apply(categorize_income)
incoming_categories = data['categorize_income'].unique()
print(incoming_categories)
vars = data['categorize_income'].value_counts()
print(vars, type(vars))

['B' 'C' 'D' 'E' 'A']
C    15921
B     5014
D      349
A       25
E       22
Name: categorize_income, dtype: int64 <class 'pandas.core.series.Series'>


In [27]:
debts_salary = ds_pivot(data, 'total_income_category', 'debt')
print(debts_salary)

                       Всего  Должников  Доля_должников
total_income_category                                  
A                         25          2        0.080000
B                       5014        354        0.070602
C                      15921       1353        0.084982
D                        349         21        0.060172
E                         22          2        0.090909


**Вывод:** 
Учитываем только категории В и С. Доля должников растёт с уменьшением доходов. К тому же они чаще обращаются за кредитами, возможно это социальный фактор.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Здесь лучше делать выводы по двум самым крупным категориям - B и С. По остальным категориям данных недостаточно для корректных выводов, статпогрешность большая
    <div class='alert alert-info'>Скорректировал вывод.</div>
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Чем выше доход, тем ниже доля должников. Что касается частоты обращения, то людей с доходом категории С может быть больше в принципе, вне зависимости от кредитов
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
debts_purpose = ds_pivot(data, 'purpose_category', 'debt')
print(debts_purpose)

                          Всего  Должников  Доля_должников
purpose_category                                          
операции с автомобилем     4279        400        0.093480
операции с недвижимостью  10751        780        0.072551
получение образования      3988        369        0.092528
проведение свадьбы         2313        183        0.079118


**Вывод:** Самая частая причина - недвижимость, она же и самая надёжная по возвратам, на свадьбы берут реже, возможно суммы меньше - поэтому возвраты лучше.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Можно выделить и наиболее рискованную категорию, а также подумать над причинами такого распределения
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
- Человеческий фактор: Возможно люди не захотели вводить данные (по доходам или семейное положение).
- Технический фактор: Пропуски наблюдаются только в типе данных float64, возможно это ошибки передачи данных от некоторых служб, отвечающих за них, или несвоевременная передача данных, участвующих в вычислениях (например данные из налоговой службы, пенсионного фонда)

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Тут важно отметить человеческий и технический факторы. Дополни, пожалуйста, причины, исходя из этого
    <div class='alert alert-info'>Скорректировал вывод.</div>
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Пропуски были только в столбцах с доходом и стажем, а не с семейным положением. Клиенты могли не предоставить данные. Возможно, эту информацию не требуют в некоторых случаях. Проблемы могли возникнуть на этапе загрузки и на этапе выгрузки данных. Неправильный запрос, объединение таблиц также может быть причиной.

Проверить, что пропуски в 'total_income' и 'days_employed' находятся на одних строках, можно, например, так: 
    
    
    len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())])
    
Значений 2174, что совпадает со значениями пропусков в обоих столбцах по отдельности.

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение "фильтрует" выбросы в количественных параметрах, тогда как среднее арифметическое, учитывая выбросы может существенно исказить картину.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Да, смотрим на наличие выбросов👍 Среднее и медиана различаются, если в выборке присутствуют выбросы. Медиана позволяет нивелировать их влияние.
</div>

### Шаг 4: общий вывод.

Возможно стоит рассмотреть совместный анализ двух(трёх) параметров. Из анализа видно, что лучше всего возвращают кредиты по сделкам с недвижимостью, семейные пары, на свадьбы, без детей. Логика возможно такая: решили пожениться и завести детей - берут кредит на свадьбу и жильё - серьёзные намерения - выше процент возвратов кредита вовремя.

**Заключение**

    Полученные данные были обработаны - выявлены аномальные значения по количеству детей и ошибки в количественных данных.
    Аномалии по детям удалены, по количественным данным заменены медианными значениями.
    Добавлена категоризация данных по уровню доходов, по семейному положению.
    
    В результате исследования выявлено:
    - С появлением и увеличением количества детей доля должников по кредитам возрастает.
    - Одинокие люди, утратившие статус замужем/женат лучше выплачивают кредиты. Хужевсего выплачивают люди вообще не имеющие статуса. Люди, побывавшие в официальном браке, наиболее надежны. Возможно, есть связь с возрастом.
    - Учитывая репрезентативность только выборки по людям с определённым доходом: доля должников растёт с уменьшением доходов. К тому же они чаще обращаются за кредитами, возможно это социальный фактор.
    - Самая частая причина  - взятия кредита недвижимость, она же и самая надёжная по возвратам, на свадьбы берут реже, но возвраты так же хорошие, возможно необходимо рассмотреть зависимость от сумм кредитов.
    
    В итоге:
    - Наиболее надёжным заёмщиком представляются люди, имеющие семейный статус, с 1 ребёнком (или без детей), берущие кредит на недвижимость.
    - Наиблее рискованным - берущий на автомобиль или образование пир низком уровне доходов.
    
    Рекомендации
    
    При решении о выдаче кредита необходимо учитывать указанные особенности. Возможно стоит провести дополнительное исследование на взаимное влияние факторов.
    
 


<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
 
Выводы нужно будет дополнить. В первую очередь обращают внимание на выводы - это отчет перед заказчиком. Здесь нужно описать какая работа над данными была проведена и каких результатов удалось добиться. Опиши все зависимости. Рекомендованная структура выводов следующая:
    
- абзац по предобработке данных, что было сделано
    
- выявленные зависимости просрочки от разных факторов в общем виде
    
- портреты надежного и рискованного заемщика
    
- рекомендации заказчику
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
В качестве рекомендаций можно предложить другую категоризацию по доходам. Можно попробовать разбить всех клиентов на равные по количеству группы. Для категоризации данных также можно использовать методы pd.cut() и pd.qcut(). Подробнее о них можно почитать здесь https://www.geeksforgeeks.org/how-to-use-pandas-cut-and-qcut/ или <a href="https://dfedorov.spb.ru/pandas/%D0%A0%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5%20%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85%20%D0%B2%20Pandas%20%D1%81%20%D0%BF%D0%BE%D0%BC%D0%BE%D1%89%D1%8C%D1%8E%20qcut%20%D0%B8%20cut.html">здесь</a>. Можно предложить исследовать другие признаки, рассчитать другие метрики (отношение дохода к сумме выплат и др.). В анкете сделать поля о доходе и стаже обязательными к заполнению, добавить автопроверку на формат вводимых пользователем данных
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
Ты хорошо постарался, но проект еще нуждается в доработке. Нужно построить сводные в каждом вопросе, проверить выводы, дополнить причины пропусков. Доделывай и присылай на повторную проверку. Дополнительно можно подумать над функцией, которую можно использовать во всех вопросах
        
Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний больше нет. Проект принят! 👏 Поздравляю и желаю удачи!
</font></b>
</div>